src/libsystemd/sd-event/sd-event.c

```diff
@@ -179,12 +201,30 @@ static int prepare_prioq_compare(const void *a, const void *b) {
         return CMP(x->priority, y->priority);
 }
 
+static usec_t time_event_source_next(const sd_event_source *s) {
+        assert(s);
+
```

Comment:
Maybe we should add `assert(s->ratelimited || EVENT_SOURCE_IS_TIME(s->type))`.

Reply:
i had that in first, but dropped it. The reason is to make it more robust to calling this in a not so carefully chosen order. i.e. with the current patch's code you can insert any event source into the time prioqs, they will be ordered to the end of it, and thus be ignored. This means we can first insert them into the prioq and then afterwards set the ratelimit boolean, and all is good. 

If we'd add the assert here, then we'd have to carefully make sure to strictly set the bool first, and only insert then.

I like code that isn't fragily towards "mild" out-of-order behaviour, if you follow what I mean.